File: src/main/java/editor/gui/generic/RepeatButton.java
package editor.gui.generic;

import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;

import javax.swing.JButton;
import javax.swing.Timer;

import editor.util.MouseListenerFactory;

/**
 * This class is a button that can repeat its action when held down after a customizable delay.
 * If the mouse is moved off the button while held down, actions will still continue until it
 * is released.  Default delays set by constructors can be overrided by using the appropriate
 * methods.
 * 
 * @author Alec Roelke
 */
@SuppressWarnings("serial")
public class RepeatButton extends JButton
{
    /**
     * Default delay before starting repetition.
     */
    private int defaultInitial;
    /**
     * Default delay between repetitions.
     */
    private int defaultTick;

    /**
     * Create a new RepeatButton with specified default delays.
     * 
     * @param initial default initial delay
     * @param tick default delay between repetitions
     */
    public RepeatButton(int initial, int tick)
    {
        super();
        defaultInitial = initial;
        defaultTick = tick;
    }

    /**
     * Create a new RepeatButton with the same default delay before and between repeating
     * actions.
     * 
     * @param delay default delay
     */
    public RepeatButton(int delay)
    {
        this(delay, delay);
    }

    /**
     * Create a new RepeatButton with no delay before and between repeating actions.
     */
    public RepeatButton()
    {
        this(0);
    }

    /**
     * Add an action that will be performed upon clicking the button, after an initial delay,
     * and repeated after another delay until the mouse is released.
     * 
     * @param initial delay after first pressing the mouse button to start repeating
     * @param tick delay between repetitions
     * @param l action to perform when pressing the button
     */
    public void addRepeatListener(int initial, int tick, ActionListener l)
    {
        Timer timer = new Timer(tick, l);
        timer.setRepeats(true);
        timer.setInitialDelay(initial);
        addMouseListener(MouseListenerFactory.createHoldListener((e) -> {
            l.actionPerformed(new ActionEvent(e.getSource(), e.getID(), e.paramString(), e.getWhen(), e.getModifiersEx()));
            timer.start();
        }, (e) -> timer.stop()));
    }

    /**
     * Add an action that will be performed upon clicking the button and repeated with a
     * specified delay as long as the button is held down.
     * 
     * @param delay delay between repetitions
     * @param l action to perform
     */
    public void addRepeatListener(int delay, ActionListener l)
    {
        addRepeatListener(delay, delay, l);
    }

    /**
     * Add an action that will be performed upon clicking the button, after an initial delay,
     * and repeated after another delay until the mouse is released. The delays are the defaults
     * set by RepeatButton's constructor.
     */
    public void addRepeatListener(ActionListener l)
    {
        addRepeatListener(defaultInitial, defaultTick, l);
    }
}

File: src/main/scala/editor/gui/generic/RepeatButton.scala
package editor.gui.generic

import editor.util.MouseListenerFactory

import java.awt.event.ActionEvent
import java.awt.event.ActionListener
import javax.swing.JButton
import javax.swing.Timer

/**
 * Button that, when held down, repeats its action with a delay until the mouse is released.
 * Repeated actions are not performed when the button is initially pressed; use [[addActionListener]]
 * inherited from [[JButton]] for that.
 * 
 * @constructor create a new button optionally wi th default delays
 * @param initial initial delay between pressing the button and when action repetition begins
 * @param tick delay between action repetitions; defaults to initial if not set
 * 
 * @author Alec Roelke
 */
class RepeatButton(initial: Int = 0, tick: Option[Int] = None) extends JButton {
  /**
   * Add a new repeat listener with custom initial delay and delays between repetitions.
   * 
   * @param listener action to perform while held down
   * @param initial delay before beginning repetition; use the default value if unset
   * @param tick delay between repetitions; use the default value if unset
   */
  def addRepeatListener(listener: ActionListener, initial: Int = initial, tick: Int = tick.getOrElse(initial)): Unit = {
    val timer = Timer(tick, listener)
    timer.setRepeats(true)
    timer.setInitialDelay(initial)
    addMouseListener(MouseListenerFactory.createHoldListener((e) => {
      listener.actionPerformed(ActionEvent(e.getSource, e.getID, e.paramString, e.getWhen, e.getModifiersEx))
      timer.start()
    }, _ => timer.stop()))
  }

  /**
   * Add a new repeat listener with the same initial delay and tick.
   * 
   * @param listener action to perform while held down
   * @param delay delay between actions and before the repetitions start
   */
  def addRepeatListener(listener: ActionListener, delay: Int): Unit = addRepeatListener(listener, delay, delay)
}